server/datastore/mysql/activities.go

```diff
@@ -261,12 +261,12 @@ func (ds *Datastore) ListHostUpcomingActivities(ctx context.Context, hostID uint
 		`SELECT
 			COUNT(*) c
 			FROM host_software_installs hsi
-			WHERE hsi.host_id = :host_id AND
+			WHERE hsi.host_id = :host_id AND hsi.software_installer_id IS NOT NULL AND
 				hsi.status = :software_status_install_pending`,
 		`SELECT
 			COUNT(*) c
 			FROM host_software_installs hsi
-			WHERE hsi.host_id = :host_id AND
+			WHERE hsi.host_id = :host_id AND hsi.software_installer_id IS NOT NULL AND
```

Comment:
This is the actual fix for the bug reported by QA wolf right? Should the other changes be linked to different issue?

Reply:
Underlying cause for the QAWolf bug is that we now need to manually clean up pending installs when installers get deleted, rather than the pending installs getting deleted automatically due to the `ON DELETE CASCADE` that was previously on `host_software_installs.software_installer_id`. This fix is belt-and-suspenders so the apparent problem goes away in QAWolf's environment without either a migration or a manual DB query, but without the larger cleanup we'd have invisible pending (un)installs when an installer is deleted, and the install part is a regression from v4.58.0.

Reply:
Basically, #23350 exists because the implementation of #21654 was incomplete, so this is unreleased whichever way we slice this.